# Assignment 3: Keras & CNNs

## Submitted by: Aarathi Vijayachandran Chettiar Bhagavathi (244267)

### Assignment question is available here: https://ovgu-ailab.github.io/idl2023/assignment3.html

## CNN on MNIST dataset

### Code - 1

In [1]:
import tensorflow as tf
from tensorflow.keras import layers, models

# Load the MNIST dataset
mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# Normalize pixel values to be in the range [0, 1]
x_train, x_test = x_train / 255.0, x_test / 255.0

# Reshape the data to have the format (width, height, channels)
x_train = x_train.reshape(x_train.shape[0], 28, 28, 1)
x_test = x_test.reshape(x_test.shape[0], 28, 28, 1)

# Create a sequential model
model = models.Sequential()

# Add Convolutional layers with MaxPool2D layers
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))

# Flatten the output for the fully connected layers
model.add(layers.Flatten())

# Add Dense (fully connected) layers
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))  # 10 classes for digits 0-9

# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# Train the model
model.fit(x_train, y_train, epochs=5, validation_data=(x_test, y_test))

# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print(f"Test accuracy: {test_acc}")


11490434/11490434 [==============================] - 1s 0us/step
Epoch 1/5
1875/1875 [==============================] - 59s 31ms/step - loss: 0.1400 - accuracy: 0.9563 - val_loss: 0.0649 - val_accuracy: 0.9772
Epoch 2/5
1875/1875 [==============================] - 55s 29ms/step - loss: 0.0465 - accuracy: 0.9856 - val_loss: 0.0432 - val_accuracy: 0.9860
Epoch 3/5
1875/1875 [==============================] - 54s 29ms/step - loss: 0.0339 - accuracy: 0.9898 - val_loss: 0.0490 - val_accuracy: 0.9846
Epoch 4/5
1875/1875 [==============================] - 55s 29ms/step - loss: 0.0261 - accuracy: 0.9919 - val_loss: 0.0315 - val_accuracy: 0.9906
Epoch 5/5
313/313 [==============================] - 4s 12ms/step - loss: 0.0324 - accuracy: 0.9915
Test accuracy: 0.9915000200271606


CNN on MNIST dataset:


*   Keras used
*   3 * 3 kernel used
*   3 Conv2D and 2 MaxPool2D layers added
*   2 Dense layers added
*   Adam optimizer used

High test accuracy obtained. Test accuracy: 0.9915000200271606

## CNN on CIFAR dataset

### Code - 2

In [5]:
import tensorflow as tf
from tensorflow.keras import layers, models
from tensorflow.keras.optimizers import Adam

# Load the CIFAR-10 dataset
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()

# Normalize pixel values to be in the range [0, 1]
x_train, x_test = x_train / 255.0, x_test / 255.0

# Convert labels to 1D array
y_train = y_train.reshape(-1)
y_test = y_test.reshape(-1)

# Create a sequential model
model = models.Sequential()

# Add Convolutional layers with MaxPool2D layers
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))

# Flatten the output for the fully connected layers
model.add(layers.Flatten())

# Add Dense (fully connected) layers
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))  # 10 classes for CIFAR-10

# Compile the model with Adam optimizer
optimizer = Adam(learning_rate=0.001)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# Train the model
model.fit(x_train, y_train, epochs=10, validation_data=(x_test, y_test))

# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print(f"Test accuracy: {test_acc}")


170498071/170498071 [==============================] - 10s 0us/step
Epoch 1/10
1563/1563 [==============================] - 71s 45ms/step - loss: 1.5310 - accuracy: 0.4380 - val_loss: 1.2512 - val_accuracy: 0.5530
Epoch 2/10
1563/1563 [==============================] - 70s 45ms/step - loss: 1.1608 - accuracy: 0.5879 - val_loss: 1.1162 - val_accuracy: 0.6059
Epoch 3/10
1563/1563 [==============================] - 68s 43ms/step - loss: 1.0080 - accuracy: 0.6456 - val_loss: 1.0276 - val_accuracy: 0.6361
Epoch 4/10
1563/1563 [==============================] - 68s 44ms/step - loss: 0.9108 - accuracy: 0.6810 - val_loss: 0.9769 - val_accuracy: 0.6597
Epoch 5/10
1563/1563 [==============================] - 68s 44ms/step - loss: 0.8345 - accuracy: 0.7076 - val_loss: 0.9040 - val_accuracy: 0.6895
Epoch 6/10
1563/1563 [==============================] - 70s 45ms/step - loss: 0.7768 - accuracy: 0.7279 - val_loss: 0.9035 - val_accuracy: 0.6900
Epoch 7/10
1563/1563 [==============================] - 

CNN on CIFAR dataset:
*   Keras used
*   3 * 3 kernel used
*   3 Conv2D and 2 MaxPool2D layers added
*   2 Dense layers added
*   Adam optimizer used

Test accuracy: 0.7064999938011169

## Experiments :

### 1) Using 5*5 kernel on Code - 2

In [13]:
import tensorflow as tf
from tensorflow.keras import layers, models
from tensorflow.keras.optimizers import Adam

# Load the CIFAR-10 dataset
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()

# Normalize pixel values to be in the range [0, 1]
x_train, x_test = x_train / 255.0, x_test / 255.0

# Convert labels to 1D array
y_train = y_train.reshape(-1)
y_test = y_test.reshape(-1)

# Create a sequential model
model = models.Sequential()

# Add Convolutional layers with MaxPool2D layers and larger filter sizes
model.add(layers.Conv2D(64, (5, 5), activation='relu', input_shape=(32, 32, 3)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(128, (5, 5), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(256, (5, 5), activation='relu'))

# Flatten the output for the fully connected layers
model.add(layers.Flatten())

# Add Dense (fully connected) layers
model.add(layers.Dense(256, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))  # 10 classes for CIFAR-10

# Compile the model with Adam optimizer
optimizer = Adam(learning_rate=0.001)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# Train the model
model.fit(x_train, y_train, epochs=10, validation_data=(x_test, y_test))

# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print(f"Test accuracy: {test_acc}")


Epoch 1/10
1563/1563 [==============================] - 301s 192ms/step - loss: 1.4696 - accuracy: 0.4666 - val_loss: 1.2487 - val_accuracy: 0.5536
Epoch 2/10
1563/1563 [==============================] - 294s 188ms/step - loss: 1.1178 - accuracy: 0.6036 - val_loss: 1.1125 - val_accuracy: 0.6076
Epoch 3/10
1563/1563 [==============================] - 297s 190ms/step - loss: 0.9446 - accuracy: 0.6662 - val_loss: 0.9982 - val_accuracy: 0.6516
Epoch 4/10
1563/1563 [==============================] - 296s 189ms/step - loss: 0.8100 - accuracy: 0.7151 - val_loss: 1.0035 - val_accuracy: 0.6588
Epoch 5/10
1563/1563 [==============================] - 293s 188ms/step - loss: 0.7055 - accuracy: 0.7502 - val_loss: 0.9345 - val_accuracy: 0.6869
Epoch 6/10
1563/1563 [==============================] - 293s 187ms/step - loss: 0.6108 - accuracy: 0.7834 - val_loss: 1.0298 - val_accuracy: 0.6743
Epoch 7/10
1563/1563 [==============================] - 283s 181ms/step - loss: 0.5271 - accuracy: 0.8129 - val_

Small reduction in accuracy from Test accuracy of Code-2: 0.7064999938011169 to Test accuracy: 0.675000011920929

### 2) Using Average Pooling instead of Max Pooling on Code - 2

In [14]:
import tensorflow as tf
from tensorflow.keras import layers, models
from tensorflow.keras.optimizers import Adam

# Load the CIFAR-10 dataset
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()

# Normalize pixel values to be in the range [0, 1]
x_train, x_test = x_train / 255.0, x_test / 255.0

# Convert labels to 1D array
y_train = y_train.reshape(-1)
y_test = y_test.reshape(-1)

# Create a sequential model
model = models.Sequential()

# Add Convolutional layers with AveragePooling2D layers
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)))
model.add(layers.AveragePooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.AveragePooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))

# Flatten the output for the fully connected layers
model.add(layers.Flatten())

# Add Dense (fully connected) layers
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))  # 10 classes for CIFAR-10

# Compile the model with Adam optimizer
optimizer = Adam(learning_rate=0.001)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# Train the model
model.fit(x_train, y_train, epochs=10, validation_data=(x_test, y_test))

# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print(f"Test accuracy: {test_acc}")


Epoch 1/10
1563/1563 [==============================] - 59s 37ms/step - loss: 1.5990 - accuracy: 0.4166 - val_loss: 1.3427 - val_accuracy: 0.5133
Epoch 2/10
1563/1563 [==============================] - 57s 36ms/step - loss: 1.2498 - accuracy: 0.5546 - val_loss: 1.1859 - val_accuracy: 0.5790
Epoch 3/10
1563/1563 [==============================] - 59s 38ms/step - loss: 1.0997 - accuracy: 0.6135 - val_loss: 1.0490 - val_accuracy: 0.6253
Epoch 4/10
1563/1563 [==============================] - 60s 38ms/step - loss: 1.0006 - accuracy: 0.6467 - val_loss: 1.0141 - val_accuracy: 0.6441
Epoch 5/10
1563/1563 [==============================] - 59s 38ms/step - loss: 0.9197 - accuracy: 0.6777 - val_loss: 0.9478 - val_accuracy: 0.6658
Epoch 6/10
1563/1563 [==============================] - 58s 37ms/step - loss: 0.8623 - accuracy: 0.6974 - val_loss: 0.9161 - val_accuracy: 0.6797
Epoch 7/10
1563/1563 [==============================] - 57s 37ms/step - loss: 0.8129 - accuracy: 0.7161 - val_loss: 0.9675 -

Changing type of pooling does not have much effect on accuracy. Test accuracy: 0.7010999917984009, it is almost same as that of Code-2

### 3) Using Strided Convolutions instead of Max Pooling on Code - 2

In [1]:
import tensorflow as tf
from tensorflow.keras import layers, models
from tensorflow.keras.optimizers import Adam

# Load the CIFAR-10 dataset
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()

# Normalize pixel values to be in the range [0, 1]
x_train, x_test = x_train / 255.0, x_test / 255.0

# Convert labels to 1D array
y_train = y_train.reshape(-1)
y_test = y_test.reshape(-1)

# Create a sequential model
model = models.Sequential()

# Add Convolutional layers with strided convolutions
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3), strides=(2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu', strides=(2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu', strides=(2, 2)))

# Flatten the output for the fully connected layers
model.add(layers.Flatten())

# Add Dense (fully connected) layers
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))  # 10 classes for CIFAR-10

# Compile the model with Adam optimizer
optimizer = Adam(learning_rate=0.001)
model.compile(optimizer=optimizer,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# Train the model
model.fit(x_train, y_train, epochs=10, validation_data=(x_test, y_test))

# Evaluate the model
test_loss, test_acc = model.evaluate(x_test, y_test)
print(f"Test accuracy: {test_acc}")


Epoch 1/10
1563/1563 [==============================] - 25s 15ms/step - loss: 1.5733 - accuracy: 0.4276 - val_loss: 1.3443 - val_accuracy: 0.5078
Epoch 2/10
1563/1563 [==============================] - 26s 17ms/step - loss: 1.2592 - accuracy: 0.5516 - val_loss: 1.2149 - val_accuracy: 0.5641
Epoch 3/10
1563/1563 [==============================] - 29s 18ms/step - loss: 1.1236 - accuracy: 0.6021 - val_loss: 1.1234 - val_accuracy: 0.5970
Epoch 4/10
1563/1563 [==============================] - 27s 17ms/step - loss: 1.0327 - accuracy: 0.6353 - val_loss: 1.1195 - val_accuracy: 0.6069
Epoch 5/10
1563/1563 [==============================] - 24s 16ms/step - loss: 0.9607 - accuracy: 0.6606 - val_loss: 1.0350 - val_accuracy: 0.6356
Epoch 6/10
1563/1563 [==============================] - 24s 16ms/step - loss: 0.8935 - accuracy: 0.6852 - val_loss: 1.0177 - val_accuracy: 0.6428
Epoch 7/10
1563/1563 [==============================] - 24s 16ms/step - loss: 0.8366 - accuracy: 0.7027 - val_loss: 0.9870 -

Small reduction in accuracy from Test accuracy of Code-2: 0.7064999938011169 to Test accuracy: 0.6622999906539917